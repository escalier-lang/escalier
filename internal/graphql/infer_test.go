package graphql

import (
	"os"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/vektah/gqlparser/v2"
	"github.com/vektah/gqlparser/v2/ast"
	"github.com/vektah/gqlparser/v2/validator/rules"
)

func loadSchema(t *testing.T) *ast.Schema {
	// Read schema.graphql from disk
	schemaBytes, err := os.ReadFile("schema.graphql")
	if err != nil {
		t.Fatalf("failed to read schema.graphql: %v", err)
	}
	schemaStr := string(schemaBytes)

	// Convert SchemaDocument into a usable schema.
	schema := gqlparser.MustLoadSchema(&ast.Source{
		Name:    "schema.graphql",
		Input:   schemaStr,
		BuiltIn: false,
	})

	return schema
}

func TestInferGraphQLQuery_GetUserAndPosts(t *testing.T) {
	schema := loadSchema(t)

	// Define a sample GraphQL query string.
	queryStr := `
		query GetUserAndPosts($userId: ID!, $postLimit: Int = 5) {
		getUser(id: $userId) {
			id
			name
			role
			contactInfo {
				email
				phone
			}
			posts {
				id
				title
				content
				comments(limit: $postLimit) {
					id
					message
					author {
						id
						name
						}
					}
				}
			}
		}
	`

	// Parse the query against the schema.
	queryDoc := gqlparser.MustLoadQueryWithRules(schema, queryStr, rules.NewDefaultRules())

	result := InferGraphQLQuery(schema, queryDoc)

	// Verify that result types are not nil
	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	// Snapshot test the inferred types
	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_CreatePostAndComment(t *testing.T) {
	schema := loadSchema(t)

	mutationStr := `
		mutation CreatePostAndComment {
			createPost(input: {
				title: "Understanding GraphQL Unions",
				content: "Unions let you return different object types from a field.",
				metadata: {
				tags: ["GraphQL", "Schema", "Unions"],
				estimatedReadTime: 5
				}
			}) {
				id
				title
				author {
					id
					name
				}
			}

			addComment(input: {
				postId: "post-456",
				message: "Great explanation of unions!"
			}) {
				id
				message
				author {
					id
					name
				}
				post {
					id
					title
				}
			}
		}
	`

	// Parse the query against the schema.
	queryDoc := gqlparser.MustLoadQueryWithRules(schema, mutationStr, rules.NewDefaultRules())

	result := InferGraphQLQuery(schema, queryDoc)

	// Verify that result types are not nil
	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	// Snapshot test the inferred types
	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_SearchContentWithUnion(t *testing.T) {
	schema := loadSchema(t)

	queryStr := `
	query SearchContent($searchTerm: String!) {
		search(text: $searchTerm) {
			... on User {
				id
				name
				role
				contactInfo {
					email
				}
			}

			... on Post {
				id
				title
				content
				author {
					name
					role
				}
			}

			... on Comment {
				id
				message
				author {
					name
				}
				post {
					title
				}
			}
		}
	}`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, queryStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}
	// Snapshot test the inferred types
	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_ScalarTypes(t *testing.T) {
	schema := loadSchema(t)

	queryStr := `
		query GetNodeWithScalars($nodeId: ID!) {
			node(id: $nodeId) {
				id
				createdAt
				... on Post {
					metadata
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, queryStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_ListTypes(t *testing.T) {
	schema := loadSchema(t)

	queryStr := `
		query GetPostsAndComments($offset: Int = 0, $limit: Int = 10, $commentLimit: Int = 5) {
			listPosts(offset: $offset, limit: $limit) {
				id
				title
				content
				comments(limit: $commentLimit) {
					id
					message
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, queryStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_InputObjectTypes(t *testing.T) {
	schema := loadSchema(t)

	mutationStr := `
		mutation CreatePostWithMetadata($postData: PostInput!) {
			createPost(input: $postData) {
				id
				title
				content
				metadata
				author {
					id
					name
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, mutationStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_SubscriptionTypes(t *testing.T) {
	schema := loadSchema(t)

	subscriptionStr := `
		subscription WatchPostCreated {
			postCreated {
				id
				title
				author {
					id
					name
					role
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, subscriptionStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_InterfaceTypes(t *testing.T) {
	schema := loadSchema(t)

	queryStr := `
		query GetNodeInterface($nodeId: ID!) {
			node(id: $nodeId) {
				id
				createdAt
				... on User {
					name
					role
					contactInfo {
						email
						phone
					}
				}
				... on Post {
					title
					content
					author {
						name
					}
				}
				... on Comment {
					message
					author {
						name
					}
					post {
						title
					}
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, queryStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_NullableFields(t *testing.T) {
	schema := loadSchema(t)

	queryStr := `
		query GetPostWithNullableFields($postId: ID!) {
			getUser(id: $postId) {
				id
				name
				contactInfo {
					email
					phone
				}
				posts {
					id
					title
					content
					metadata
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, queryStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_ComplexMutation(t *testing.T) {
	schema := loadSchema(t)

	mutationStr := `
		mutation UpdateUserAndCreatePost($userId: ID!, $newRole: Role!, $postInput: PostInput!) {
			updateUserRole(userId: $userId, role: $newRole) {
				id
				name
				role
				contactInfo {
					email
					phone
				}
			}
			createPost(input: $postInput) {
				id
				title
				content
				metadata
				author {
					id
					name
					role
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, mutationStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}

func TestInferGraphQLQuery_SubscriptionWithVariables(t *testing.T) {
	schema := loadSchema(t)

	subscriptionStr := `
		subscription WatchComments($postId: ID!) {
			commentAdded(postId: $postId) {
				id
				message
				author {
					id
					name
					role
				}
				post {
					id
					title
				}
			}
		}
	`

	queryDoc := gqlparser.MustLoadQueryWithRules(schema, subscriptionStr, rules.NewDefaultRules())
	result := InferGraphQLQuery(schema, queryDoc)

	if result.ResultType == nil {
		t.Error("Expected ResultType to be non-nil")
	}
	if result.VariablesType == nil {
		t.Error("Expected VariablesType to be non-nil")
	}

	snaps.MatchSnapshot(t, result.ResultType.String())
	snaps.MatchSnapshot(t, result.VariablesType.String())
}
