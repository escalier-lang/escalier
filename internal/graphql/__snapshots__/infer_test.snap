
[TestInferGraphQLQuery_GetUserAndPosts - 1]
{getUser?: {id: ID, name: string, role: "ADMIN" | "USER" | "GUEST", contactInfo: {email: string, phone?: string | null}, posts: Array<{id: ID, title: string, content?: string | null, comments: Array<{id: ID, message: string, author: {id: ID, name: string}}>}>} | null}
---

[TestInferGraphQLQuery_GetUserAndPosts - 2]
{userId: ID, postLimit?: number | null}
---

[TestInferGraphQLQuery_CreatePostAndComment - 1]
{createPost: {id: ID, title: string, author: {id: ID, name: string}}, addComment: {id: ID, message: string, author: {id: ID, name: string}, post: {id: ID, title: string}}}
---

[TestInferGraphQLQuery_CreatePostAndComment - 2]
{}
---

[TestInferGraphQLQuery_SearchContentWithUnion - 1]
{search: Array<{id: ID, name: string, role: "ADMIN" | "USER" | "GUEST", contactInfo: {email: string}} | {id: ID, title: string, content?: string | null, author: {name: string, role: "ADMIN" | "USER" | "GUEST"}} | {id: ID, message: string, author: {name: string}, post: {title: string}}>}
---

[TestInferGraphQLQuery_SearchContentWithUnion - 2]
{searchTerm: string}
---

[TestInferGraphQLQuery_ScalarTypes - 1]
{node?: Node | null}
---

[TestInferGraphQLQuery_ScalarTypes - 2]
{nodeId: ID}
---

[TestInferGraphQLQuery_ListTypes - 1]
{listPosts: Array<{id: ID, title: string, content?: string | null, comments: Array<{id: ID, message: string}>}>}
---

[TestInferGraphQLQuery_ListTypes - 2]
{offset?: number | null, limit?: number | null, commentLimit?: number | null}
---

[TestInferGraphQLQuery_InputObjectTypes - 1]
{createPost: {id: ID, title: string, content?: string | null, metadata?: JSON | null, author: {id: ID, name: string}}}
---

[TestInferGraphQLQuery_InputObjectTypes - 2]
{postData: {title: string, content?: string | null, metadata?: JSON | null}}
---

[TestInferGraphQLQuery_SubscriptionTypes - 1]
{postCreated: {id: ID, title: string, author: {id: ID, name: string, role: "ADMIN" | "USER" | "GUEST"}}}
---

[TestInferGraphQLQuery_SubscriptionTypes - 2]
{}
---

[TestInferGraphQLQuery_InterfaceTypes - 1]
{node?: Node | null}
---

[TestInferGraphQLQuery_InterfaceTypes - 2]
{nodeId: ID}
---

[TestInferGraphQLQuery_NullableFields - 1]
{getUser?: {id: ID, name: string, contactInfo: {email: string, phone?: string | null}, posts: Array<{id: ID, title: string, content?: string | null, metadata?: JSON | null}>} | null}
---

[TestInferGraphQLQuery_NullableFields - 2]
{postId: ID}
---

[TestInferGraphQLQuery_ComplexMutation - 1]
{updateUserRole: {id: ID, name: string, role: "ADMIN" | "USER" | "GUEST", contactInfo: {email: string, phone?: string | null}}, createPost: {id: ID, title: string, content?: string | null, metadata?: JSON | null, author: {id: ID, name: string, role: "ADMIN" | "USER" | "GUEST"}}}
---

[TestInferGraphQLQuery_ComplexMutation - 2]
{userId: ID, newRole: "ADMIN" | "USER" | "GUEST", postInput: {title: string, content?: string | null, metadata?: JSON | null}}
---

[TestInferGraphQLQuery_SubscriptionWithVariables - 1]
{commentAdded: {id: ID, message: string, author: {id: ID, name: string, role: "ADMIN" | "USER" | "GUEST"}, post: {id: ID, title: string}}}
---

[TestInferGraphQLQuery_SubscriptionWithVariables - 2]
{postId: ID}
---
